omnigibson/systems/system_base.py

```diff
@@ -119,9 +220,206 @@ def get_active_systems(cls):
+                - (n, 4)-array: per-particle (x,y,z,w) quaternion orientation
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def get_particle_local_pose(cls, idx):
+        """
+        Compute particle's position and orientation in the particle's parent frame
+
+        Args:
+            idx (int): Index of the particle to compute position and orientation for. Note: this is
+                equivalent to grabbing the corresponding idx'th entry from @get_particles_local_pose()
+
+        Returns:
+            2-tuple:
+                - 3-array: (x,y,z) position
+                - 4-array: (x,y,z,w) quaternion orientation
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def set_particles_local_pose(cls, positions=None, orientations=None):
+        """
+        Sets all particles' positions and orientations that belong to this system in the particles' parent frames
+
+        Args:
+            positions (n-array): (n, 3) per-particle (x,y,z) position
+            orientations (n-array): (n, 4) per-particle (x,y,z,w) quaternion orientation
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def set_particle_local_pose(cls, idx, position=None, orientation=None):
+        """
+        Sets particle's position and orientation in the particle's parent frame
+
+        Args:
+            idx (int): Index of the particle to set position and orientation for. Note: this is
+                equivalent to setting the corresponding idx'th entry from @set_particles_local_pose()
+            position (3-array): particle (x,y,z) position
+            orientation (4-array): particle (x,y,z,w) quaternion orientation
+        """
+        raise NotImplementedError()
+
     def __init__(self):
         raise ValueError("System classes should not be created!")
 
+    @classproperty
+    def state_size(cls):
+        # We have n_particles (1), min / max scale (3*2), each particle pose (7*n)
+        return 7 + 7 * cls.n_particles
+
+    @classmethod
+    def _dump_state(cls):
+        positions, orientations = cls.get_particles_local_pose() if \
+            cls._store_local_poses else cls.get_particles_position_orientation()
+        return dict(
+            n_particles=cls.n_particles,
+            min_scale=cls.min_scale,
+            max_scale=cls.max_scale,
+            positions=positions,
+            orientations=orientations,
+        )
+
+    @classmethod
+    def _load_state(cls, state):
+        # Sanity check loading particles
+        assert cls.n_particles == state["n_particles"], f"Inconsistent number of particles found when loading " \
+                                                        f"particles state! Current number: {cls.n_particles}, " \
+                                                        f"loaded number: {state['n_particles']}"
+        # Load scale
+        cls.set_scale_limits(minimum=state["min_scale"], maximum=state["max_scale"])
```

Comment:
why do we need `set_scale_limits` at all? Can we just do 
```
cls.min_scale = state["min_scale"]
cls.max_scale = state["max_scale"]
```
Also, in load/dump_state, do we also need to handle `cls._group_scales`? Update: maybe it's handled by create/remove_attachment_group already.

Reply:
done